drivers/kscan/kscan_gpio.c

```diff
@@ -0,0 +1,320 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_kscan
+
+#include <device.h>
+#include <drivers/kscan.h>
+#include <drivers/gpio.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(kscan_gpio, CONFIG_KSCAN_LOG_LEVEL);
+
+#define MATRIX_NODE_ID DT_DRV_INST(0)
+#define MATRIX_ROWS DT_PROP_LEN(MATRIX_NODE_ID, row_gpios)
+#define MATRIX_COLS DT_PROP_LEN(MATRIX_NODE_ID, col_gpios)
+
+struct kscan_gpio_scan_item {
+	struct device *device;
+	gpio_pin_t pin;
+	gpio_flags_t flags;
+};
+
+struct kscan_gpio_config {
+};
+
+struct kscan_gpio_data {
+	struct k_timer timer;
+	kscan_callback_t callback;
+	struct k_work work;
+	bool matrix_state[MATRIX_ROWS][MATRIX_COLS];
+	struct kscan_gpio_scan_item rows[MATRIX_ROWS];
+	struct kscan_gpio_scan_item cols[MATRIX_COLS];
+	struct device *dev;
+};
+
+#if DT_ENUM_IDX(MATRIX_NODE_ID, diode_direction) == 0
+static int kscan_gpio_read(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+
+	static bool read_state[MATRIX_ROWS][MATRIX_COLS];
+
+	for (int r = 0; r < MATRIX_ROWS; r++) {
+		struct kscan_gpio_scan_item row = data->rows[r];
+
+		gpio_pin_set(row.device, row.pin, 1);
+		for (int c = 0; c < MATRIX_COLS; c++) {
+			struct kscan_gpio_scan_item col = data->cols[c];
+
+			read_state[r][c] =
+				gpio_pin_get(col.device, col.pin) > 0;
+		}
+
+		gpio_pin_set(row.device, row.pin, 0);
+	}
+
+	for (int r = 0; r < MATRIX_ROWS; r++) {
+		for (int c = 0; c < MATRIX_COLS; c++) {
+			bool pressed = read_state[r][c];
+
+			if (pressed != data->matrix_state[r][c]) {
+				data->matrix_state[r][c] = pressed;
+				data->callback(dev, r, c, pressed);
+			}
+		}
+	}
+
+	return 0;
+}
+
+#elif DT_ENUM_IDX(MATRIX_NODE_ID, diode_direction) == 1
+
+static int kscan_gpio_read(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+	static bool read_state[MATRIX_ROWS][MATRIX_COLS];
+
+	for (int c = 0; c < MATRIX_COLS; c++) {
+		struct kscan_gpio_scan_item col = data->cols[c];
+
+		gpio_pin_set(col.device, col.pin, 1);
+
+		for (int r = 0; r < MATRIX_ROWS; r++) {
+			struct kscan_gpio_scan_item row = data->rows[r];
+
+			read_state[r][c] =
+				gpio_pin_get(row.device, row.pin) > 0;
+		}
+
+		gpio_pin_set(col.device, col.pin, 0);
+	}
+
+	for (int r = 0; r < MATRIX_ROWS; r++) {
+		for (int c = 0; c < MATRIX_COLS; c++) {
+			bool pressed = read_state[r][c];
+
+			if (pressed != data->matrix_state[r][c]) {
+				data->matrix_state[r][c] = pressed;
+				data->callback(dev, r, c, pressed);
+			}
+		}
+	}
+
+	return 0;
+}
+
+#endif
+
+static void kscan_gpio_timer_handler(struct k_timer *timer)
+{
+	struct kscan_gpio_data *data =
+		CONTAINER_OF(timer, struct kscan_gpio_data, timer);
+
+	k_work_submit(&data->work);
+}
+
+static void kscan_gpio_work_handler(struct k_work *work)
+{
+	struct kscan_gpio_data *data =
+		CONTAINER_OF(work, struct kscan_gpio_data, work);
+
+	kscan_gpio_read(data->dev);
+}
+
+#define INIT_ROW_SCAN_ITEM(idx) \
```

Comment:
Except that it does not follow coding style and is not beautiful, why can you not write it as a function?

Reply:
@jfischer-phytec-iot Maybe I'm completely missing something, but see my comments on #24536 about needing to loop from 0 to max rows/columns, and generate entries in an array. In doing so, you need to get the pin/flags from the defines, so you can't loop at *runtime* to fetch it, you need to "unroll the loop" at *compile* time, AFAICT.

If I'm missing something, by all means I'd love to understand how to achieve pulling a dynamic length phandles-array out, with full details, to configure/use.

Reply:
@jfischer-phytec-iot You can see a branch I made of this work that adds a macro that eliminates the need for this ugly hack in my comment here: https://github.com/zephyrproject-rtos/zephyr/issues/24536#issuecomment-629571341